soc/xtensa/intel_adsp/ace/power.c

```diff
@@ -373,4 +385,35 @@ void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
 	z_xt_ints_on(core_desc[cpu].intenable);
 }
 
+#if CONFIG_ADSP_DYNAMIC_CLOCK_GATING
+
+atomic_t clock_gating_lock;
+
+void adsp_clock_gating_lock_get(void)
+{
+	if (atomic_get(&clock_gating_lock) == 0)
+		DSPCS.bootctl[arch_proc_id()].bctl |= DSPBR_BCTL_WAITIPCG;
+
+	atomic_inc(&clock_gating_lock);
```

Comment:
For implementation, use of atomics is not bringing much added value here.  It does seem to be harmless and functionality will be ok still.

Reply:
I don't see any other synchronization, so these are needed.  It's a reference count that might be modified from racing contexts, so you can't do it with a non-atomic load/modify/store sequence.